Sensitivty analysis workflow and functions

MatthVeron opened the pull request:
- A [script](https://github.com/JaneSullivan-NOAA/shortspine_thornyhead_2023/blob/Sensitivty_analysis/R/utils/Set_Up_Sensitivity_Analysis_workflow.R) to set up the sensitivity analysis workflow
- [Functions](https://github.com/JaneSullivan-NOAA/shortspine_thornyhead_2023/blob/Sensitivty_analysis/R/utils/sensistivity_analysis_utils.R) needed when setting up a new sensitivity analysis
- Outputs of setting up a new sensitivity analysis
   * [SA_info.RData](https://github.com/JaneSullivan-NOAA/shortspine_thornyhead_2023/blob/Sensitivty_analysis/R/Sensitivity_Anal/SA_info.RData) (hold information about all sensitivity analyses already set up)
   * [Summary](https://github.com/JaneSullivan-NOAA/shortspine_thornyhead_2023/blob/Sensitivty_analysis/model/Sensitivity_Anal/Sensitivity_analysis.pdf) of all sensitivity analyses done (info from the SA_info.RData file)
- A [routine script](https://github.com/JaneSullivan-NOAA/shortspine_thornyhead_2023/blob/Sensitivty_analysis/R/Sensitivity_Anal/Create_New_Sensitivity_Analysis.R) to set up a new sensitivity analysis (with an example)

MatthVeron commented:
Few additional things - I'm also planning to set up:
- a function to store the path for each model i.e. as soon as you create a new sensitivity analysis, the path will be saved in a text file so we keep track of both the model name and the associated directory
- a function that automatically copy and paste the input files from the base model directory to the directory of the sensitivity analysis you set up

Hopefully this kind of things fit what you guys were thinking?

JaneSullivan-NOAA commented:
@MatthVeron This is fantastic!! I can't wait to give it a try. Maybe you could give a demo next week?

Ovec8hkin commented on R/utils/sensistivity_analysis_utils.R:
```diff
@@ -0,0 +1,305 @@
+#' 3. Update the "Sensitivity_analysis.pdf" file housed in the root folder.
+#
+
+NewSensAnal <- function(topic = NULL,
```
Can we add some error checking to this function so that if a sensitivity analysis with the same properties has already been created (ie. it exists in the sensitivities data frame AND the model directory is present), the user is notified and nothing happens?

I think this would be useful so that all sensitivity scripts can start with a call to this function to setup the sensitivity, and then use all of the necessary R code to modify the relevant SS data and control files.

MatthVeron replied:
Thse checking error are  in the "Check the input arguments" section of the function I think. If this kind of things happen, the fuction stops and the user gets a notification about what is wrong (e.g. the model directory already exists). However, I'll add some checks (and options) so we can set up directories for diffderent models in the same sensitivity analysis.

I agree with you (I thik that was the idea), it would be great to start all sensitivity by calling this function and then build the R code to modify the model input files.

Ovec8hkin replied:
You're right, missed that check somehow when I was reviewing it.

Ovec8hkin reviewed (changes requested):
This generally looks good, but could you provide a fully worked example of how you envision the sensitivity workflow going using these functions? Just as an example, maybe use this workflow to run the 3.30.21 transition model without the fixed catchability parameters?

I really like the ideas behind all of this (writing a text file with information about the sensitivity to the directory is FANTASTIC, as is automatically updating a PDF file of sensitivity runs), I just think I need a concrete example to make sure I understand how this would work in practice.

MatthVeron commented:
@JaneSullivan-NOAA Thanks Jane. Of course, I'll give a demo next week. Actually I was wondering if we could set up a meeting with the whole class because I implemented the same for Rex (easier :P). So this might be Monday at the end of the GitHub training (if people can join) or at another time in the week.

@Ovec8hkin I'll build an example based on the sensitivity for the transition model (that was the plan - sorry hadn't time yesterday) - I'll do it today (on the same branch) so if you guys can have a look at it before merging the branch that would be awesome.

MatthVeron commented:
I update the workflow (a lot of new things I think). 
Important features:
- Added 'transition' as a potential topic for sensitivity analysis
- Added the possibility to consider multiple models for a given SA
- Modified the workflow to keep track of the SA ID and collapse rows when multiple models are used for a given SA (update the Summary_Sensitivity_analysis.pdf file)
- Created the update_Models_SA_table() function to generate a .pdf and dataframe to keep track of model names, directories,... for each SA
- Added multiple Readline asking the user to provide multiple details through the Rconsol to fill in the .pdf and data frame the description of the SA, models used,...

=> Will implement tomorrow the sensitivity analysis for the transition

Ovec8hkin commented:
@MatthVeron pinged you some question on GChat

Ovec8hkin reviewed (approved):
This look GREAT @MatthVeron! I am going to add a few things before merging, but this runs and serves as good example.

MatthVeron commented:
Cool !! What kind of things are you thinking to add? (I'll develop an update_SA() function to add models if needed and also wanted to track the SS version.

Ovec8hkin commented:
I just turned on console output for the SS model so users can watch the model run and I fixed a pathing problem I encountered. 

I think its be useful to add a function to get the model directory associated with a sensitivity by name (ie. 23.sq.fix), and maybe a larger wrapper to automatically plot models upon completion, and/or to handle the model comparison plotting process.

MatthVeron commented:
1- yeah I was thinking to put that as an option because sometimes it's really helpful to get the console output as a txt file to keep a track of it. I'll set it as an option.
 Sorry this file.path() is the one used for Rex... 

2- the directories of each model for each SA are already saved in the SA_info list (with another object that describes each SA). Do you think we need another object? 
I agree for the plotting stuff, I was thinking developing other functions to analyze likelihood, gradients and other stuff but didn't have time. Will try to do it next week

Ovec8hkin commented:
Yup, they're all saved, but you build up all of the directory names from scratch when doing the comparison analysis. I think it'd be better if we could just do something like `get.model.dir(model.name)` and let R go retrieve the directory from the SA_info list.

Ovec8hkin commented:
@MatthVeron Is there anything pending that you want to add, or should I merge this into main?

MatthVeron commented:
@Ovec8hkin You can close it I think, I'll ad the `get.model.dir(model.name)` stuff next week - I won't have time today nor this weekend. I think we can go through the workflow without it for now... so you're welcome to merge ;)
